format/spec.md

```diff
@@ -182,6 +182,21 @@ A **`list`** is a collection of values with some element type. The element field
 
 A **`map`** is a collection of key-value pairs with a key type and a value type. Both the key field and value field each have an integer id that is unique in the table schema. Map keys are required and map values can be either optional or required. Both map keys and map values may be any type, including nested types.
 
+#### Semi-structured Types
+
+A **`variant`** is a value that stores semi-structured data. The structure and data types in a variant are not necessarily consistent across rows in a table or data file. The variant type and binary encoding are defined in the [Parquet project](https://github.com/apache/parquet-format/blob/4f208158dba80ff4bff4afaa4441d7270103dff6/VariantEncoding.md). Support for Variant is added in Iceberg v3.
+
+Variants are similar to JSON with a wider set of primitive values including date, timestamp, timestamptz, binary, and floating points.
+
+Variant values may contain nested types:
+1. An array is an ordered collection of variant values.
+2. An object is a collection of fields that are a string key and a variant value.
+
+As a semi-structured type, there are important differences between variant and Iceberg's other types:
+1. Variant arrays are similar to lists, but may contain any variant value rather than a fixed element type.
+2. Variant objects are similar to structs, but may contain variable fields identified by name and field values may be any variant value rather than a fixed field type.
+3. Variant primitives are narrower than Iceberg's primitive types: time, timestamp_ns, timestamptz_ns, uuid, and fixed(L) are not supported.
```

Comment:
yes, mostly 464.  Is there a reason not to support fixed(L)?  I suppose it is redundant with string?

Reply:
Yes. We can use string instead. Also, we need to think of how to represent fixed(L) of length L if we want to support it. We need to encode L in the type if we don't lose such information while our type field only has 5 bits.

Reply:
I think for Fixed(L) using the same representation of String works (there can be multiple L encoded in a variant), so `size == L` this should give enough flexibility (so the type would just be "Fixed") I suppose one could get more complex then this but I'm not sure there is value given how things are currently modelled.

Reply:
> I'm assuming that for regular fixed(L) data type, when we insert a string, then it will truncate to length L if it's too long.

Yes, i think this might be a different assumptions. My assumption is that engines pass in valid `Fixed(L)` (i.e. length of this type are always exactly L).  Their would be  no truncation should happen at the storage layer.

So the encoding would be `<value header fixed><L as int32 (maybe int16 is sufficient)><string value>`

Reply:
Yeah. I think that works.  Even with truncation, that should happen on the engine side. 

We can consider adding that to the spec if there is a need.